File: CPP/059_2.cpp
#include <iostream>
#include <cstdint>
#include <vector>

#include <boost/range/algorithm/sort.hpp>
#include <boost/range/algorithm/fill.hpp>

using namespace std;

int main() {
	uint N, M, Q;
	cin >> N >> M >> Q;

	vector<pair<uint, uint>> edges(M);

	for (auto &e : edges) {
		uint x, y;
		cin >> x >> y;
		e = make_pair(x - 1, y - 1);
	}

	boost::sort(edges);

	vector<uint64_t> bits(N);
	vector<uint> goals;
	goals.reserve(64);

	while (Q > 0) {
		boost::fill(bits, 0UL);
		goals.clear();

		uint64_t bit = 1UL;
		while (bit > 0 && Q > 0) {
			uint a, b;
			cin >> a >> b;
			bits[a - 1] |= bit;
			goals.push_back(b - 1);

			bit <<= 1;
			Q--;
		}

		for (auto [a, b] : edges) {
			bits[b] |= bits[a];
		}

		bit = 1UL;
		for (uint goal : goals) {
			cout << (bits[goal] & bit ? "Yes" : "No") << endl;
			bit <<= 1;
		}
	}

	return 0;
}

File: CSharp/059_2.cs
#nullable enable

using System;
using System.Collections.Generic;

using static System.Console;
using static System.Linq.Enumerable;

using static Scanner;
using static Utils;

static class Scanner
{
	private static string[]? s;
	private static int i;

	public static string NextToken()
	{
		if (s == null || i == s.Length) {
			s = ReadLine()!.Split();
			i = 0;
		}

		return s[i++];
	}

	public static int NextInt() => int.Parse(NextToken());
	public static long NextLong() => long.Parse(NextToken());
	public static double NextDouble() => double.Parse(NextToken());
}

static class Utils
{
	public static T[] ReadArray<T>(Func<string, T> selector)
		=> ReadLine()!.Split().Select(selector).ToArray();

	public static T[] MakeArray<T>(int length, Func<int, T> selector)
		=> Range(0, length).Select(selector).ToArray();
}

public class Program
{
	public static void Main()
	{
		int N = NextInt();
		int M = NextInt();
		int Q = NextInt();

		(int, int)[] edges = MakeArray(M, i => (NextInt() - 1, NextInt() - 1));

		Array.Sort(edges);

		var bits = new ulong[N];
		var goals = new List<int>(64);

		while (Q > 0) {
			Array.Fill(bits, 0UL);
			goals.Clear();

			ulong bit = 1UL;
			while (bit > 0 && Q > 0) {
				int a = NextInt() - 1;
				int b = NextInt() - 1;
				bits[a] |= bit;
				goals.Add(b);

				bit <<= 1;
				Q--;
			}

			foreach (var (a, b) in edges) {
				bits[b] |= bits[a];
			}

			bit = 1;
			foreach (int goal in goals) {
				WriteLine((bits[goal] & bit) != 0 ? "Yes" : "No");
				bit <<= 1;
			}
		}
	}
}
